educe/ptb/head_finder.py

```diff
@@ -0,0 +1,150 @@
+    for lbl in priority_list:
+        for c_idx, (cnt, hw) in cands:
+            if cnt == lbl:
+                return (c_idx, hw)
+    else:
+        c_idx, (cnt, hw) = cands[0]
+        return (c_idx, hw)
+
+
+def _find_head_np(cnt_hws):
+    """Find head word in NP following specific rules"""
+    cands = list(enumerate(cnt_hws))
+    # return last word if tagged 'POS'
+    c_idx, (cnt, hw) = cands[-1]
+    if cnt == 'POS':
+        return (c_idx, hw)
+    # else: RL search for NN, NNP, NNPS, NNS, NX, POS or JJR
+    lset = set(['NN', 'NNP', 'NNPS', 'NNS', 'NX', 'POS', 'JJR'])
+    for c_idx, (cnt, hw) in reversed(cands):
+        if cnt in lset:
+            return (c_idx, hw)
+    # else: LR search for NP
+    for c_idx, (cnt, hw) in cands:
+        if cnt == 'NP':
+            return (c_idx, hw)
+    # else: RL search for $, ADJP or PRN
+    lset = set(['$', 'ADJP', 'PRN'])
+    for c_idx, (cnt, hw) in reversed(cands):
+        if cnt in lset:
+            return (c_idx, hw)
+    # else: RL search for CD
+    for c_idx, (cnt, hw) in reversed(cands):
+        if cnt == 'CD':
+            return (c_idx, hw)
+    # else RL search for JJ, JJS, RB, QP
+    lset = set(['JJ', 'JJS', 'RB', 'QP'])
+    for c_idx, (cnt, hw) in reversed(cands):
+        if cnt in lset:
+            return (c_idx, hw)
+    # else return last word
+    c_idx, (cnt, hw) = cands[-1]
+    return (c_idx, hw)
+
+
+def find_lexical_heads(tree):
+    """Find the lexical head at each node of a PTB tree."""
+
+    # result
+    head_word = {}  # dict(treepos, word)
+
+    # recursive helper
+    def _find_lexical_head_rec(treepos):
+        """Get the label and head word of subtree"""
+        subtree = tree[treepos]
+
+        if isinstance(subtree, Tree):
+            p_nt = subtree.label()  # parent non-terminal
+            # non-terminal and head word for each child
+            c_tposs = [tuple(list(treepos) + [c_idx])
+                       for c_idx, c in enumerate(subtree)]
+            cnt_hws = [_find_lexical_head_rec(c_tpos) for c_tpos in c_tposs]
+
+            # no head rule for unary productions
+            if len(subtree) == 1:
+                c_idx = 0
+                cnt, hw = cnt_hws[0]
+            else:
+                # use the head rule to get the head word from the children
+                if p_nt == 'NP':
+                    c_idx, hw = _find_head_np(cnt_hws)
+                else:
+                    try:
+                        drctn, prrty_lst = HEAD_RULES[p_nt]
+                    except KeyError:
+                        print(subtree)
+                        raise
+                    c_idx, hw = _find_head_generic(drctn, prrty_lst, cnt_hws)
+
+                # apply special post-rule for coordinated phrases (if needed)
+                # if h > 2 and Y_h-1 == 'CC': head = Y_h-2
+                if (c_idx > 1 and cnt_hws[c_idx - 1] == 'CC'):
+                    c_idx = c_idx - 2
+                    hw = cnt_hws[c_idx - 2]
```

Comment:
This might be my very first catch! OK, merging and gathering (with dev).
Scores to follow in two hours.


On Fri, 3 Apr 2015 at 10:36 Mathieu Morey <notifications@github.com> wrote:

> In educe/ptb/head_finder.py
> <https://github.com/kowey/educe/pull/34#discussion_r27720704>:
>
> > +                # use the head rule to get the head word from the children
> > +                if p_nt == 'NP':
> > +                    c_idx, hw = _find_head_np(cnt_hws)
> > +                else:
> > +                    try:
> > +                        drctn, prrty_lst = HEAD_RULES[p_nt]
> > +                    except KeyError:
> > +                        print(subtree)
> > +                        raise
> > +                    c_idx, hw = _find_head_generic(drctn, prrty_lst, cnt_hws)
> > +
> > +                # apply special post-rule for coordinated phrases (if needed)
> > +                # if h > 2 and Y_h-1 == 'CC': head = Y_h-2
> > +                if (c_idx > 1 and cnt_hws[c_idx - 1] == 'CC'):
> > +                    c_idx = c_idx - 2
> > +                    hw = cnt_hws[c_idx - 2]
>
> Congratulations, I think you spotted a bug :-)
> Will fix it shortly.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kowey/educe/pull/34/files#r27720704>.
>